design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,355 @@
+I could define the set of requirements that I would want to see in a new options implementation:
+
+* Options are a predefined list of possible product customizations. This list belongs to the product exclusively; it is measurable, and limited only by the business requirements. So it always easy to return even a complete set of options for rendering PDP initially. 
+
+* Variants are a unique intersection of one or may option values. The variant matrix does not belong to a product as property. The variant matrix is stored managed separately from products. The primary role of variants is to distinguish the possible intersections of options from impossible.
+The variants matrix is never meant to be returned to the storefront as is. The variants matrix will be used for filtering options allowed at the storefront after one or many options selected.
+
+### Variants, Prices & Inventory
+Variants could link the options intersection with a product, a price, or an inventory record but any of these relations is not mandatory.
+So, a variant can request data from the different domains if such data assigned.
+Such a decision brings us unseen before the level of the flexibility, since there is no difference between product price and variant price, and as the consequence variant price, could be included in B2B pricing.
+
+We do not have such behaviors either Magento 1 or 2 and as a result, the whole layer of product types such as bundle and downloadable do not support B2B prices or special prices for options.
+
+The link on a product that exists in the same domain could be done through the shared ID, which as for me makes sense since I can see a possible scenario when variation that initially had only a price will be "promoted" to a product. For example due to integration reasons, to pass additional attributes data along with variant to Google Merchant Center.
+
+### Enhanced option values
+
+To reach a better user experience,
+the option could be extended with an image that represents it 
+or info URL that provides additional information about option value.
+
+[Magento: Swatches](https://docs.magento.com/user-guide/catalog/swatches.html)
+
+[Magento: Downloadable products, options with samples](https://docs.magento.com/user-guide/catalog/product-create-downloadable.html)
+
+Both types of resources could be specified as attributes of `ProductOptionValue`.
+
+```proto
+syntax = "proto3";
+
+message Product {
+    repeated ProductOption options = 100;
+}
+
+message ProductOptionValue {
+    string id = 1;
+    string label = 2;
+    string sortOrder = 3;
+    string isDefault = 4;
+    string imageUrl = 5;
+    string infoUrl = 6;
+}
+
+message ProductOption {
+    string id = 1;
+    string label = 2;
+    string sortOrder = 3;
+    string isRequired = 4;
+    string renderType = 6;
+    repeated ProductOptionValue values = 5;
+}
+
+message ProductVariant {
+    repeated string optionValueId = 1;
```

Comment:
Let's try to think about some pizza examples:

We are Dominos. No stock as we can produce all pizza in the world, so variation represents a meaningful price for the pizza. There might be few options: size, style, topping and cheese levels which affect prices. Also, there are other options such as `cut`(pie cut, square cut, no cut) which don't affect the price. Also, there is a constraint on the `styles` based on the `size` of the pizza, these constraints do not affect prices. E.g. `small` -> `hand tossed`, `gluten free` only; `medium` - `hand tossed`, `crunchy thin crust`, `handmade pan`; `large` -> `brooklyn style`, `hand tossed`, `crunchy thin crust`;

- How to define/return optional options like `cut`?
- How to define that `brooklyn style` is only for `large` pizzas? Remember? They have the same price and no stock so it's reasonable to use one variation.
- Sometimes there will be a need to defined one variation in multiple weights. E.g. the same price for no `sauce` option, if you select sauce option you need to select an additional `sauce type` option:
<img width="662" alt="Screenshot 2020-08-04 at 10 43 02" src="https://user-images.githubusercontent.com/1028034/89314692-96d44980-d63f-11ea-85d6-7d9da526f656.png">
<img width="767" alt="Screenshot 2020-08-04 at 10 42 50" src="https://user-images.githubusercontent.com/1028034/89314723-9fc51b00-d63f-11ea-8649-e98828ff467b.png">

- The `additional cheese` option has the same price for all large pizzas, but large pizzas themself may have different prices. How to define `additional cheese` variation in a way that will allow to reuse it in different pizzas?
- Next level: `Additional cheese cannot be removed from Dominos Handmade Pan pizzas due to product quality`. So,  the common non-required option is actually required for one selected pizza style. How to defined such things in addition to the previous point?
<img width="595" alt="Screenshot 2020-08-04 at 10 51 48" src="https://user-images.githubusercontent.com/1028034/89315505-896b8f00-d640-11ea-9015-769c2fd69660.png">


Other notes:
- Having a weight attribute in this interface looks like a leaky abstraction. Is it possible to hide it, so api interfaces will be easier to work with?
- The current interface requires to describe all possible combinations. However, we know that some

Reply:
This behavior does not support right now by Magento, and the document does not cover this explicitly. Despite this fact, such behavior could be easaly achieved for the future by adding a property at the productOptionValue, which will restrict other productOptionValues to be shown. 

I removed `weight` attribute, agree with you it was a leak of implementation details that should not be exposed.